File: python/vapor_idx/indexes/range_.py
# =============================================================================
# vapor-idx — indexes/range_.py
# Sorted numeric index using Python's bisect module.
# Supports: gt, lt, gte, lte
# =============================================================================

from __future__ import annotations
import bisect
from typing import Any


class RangeIndex:
    def __init__(self) -> None:
        # field → list of (value, id) sorted ascending
        self._index: dict[str, list[tuple[float, str]]] = {}

    # ── Mutation ───────────────────────────────────────────────────────────────

    def add(self, field: str, value: Any, record_id: str) -> None:
        if value is None:
            return
        raw_values = value if isinstance(value, list) else [value]
        entries = self._index.setdefault(field, [])
        for raw in raw_values:
            num = _to_number(raw)
            if num is None:
                continue
            # bisect_left on (value, id) tuples — tuples compare lexicographically
            pos = bisect.bisect_left(entries, (num, record_id))
            entries.insert(pos, (num, record_id))

    def remove(self, field: str, value: Any, record_id: str) -> None:
        if value is None:
            return
        entries = self._index.get(field)
        if not entries:
            return
        raw_values = value if isinstance(value, list) else [value]
        for raw in raw_values:
            num = _to_number(raw)
            if num is None:
                continue
            pos = bisect.bisect_left(entries, (num, record_id))
            if pos < len(entries) and entries[pos] == (num, record_id):
                entries.pop(pos)
        if not entries:
            del self._index[field]

    # ── Lookup ─────────────────────────────────────────────────────────────────

    def gt(self, field: str, threshold: float) -> set[str]:
        entries = self._index.get(field, [])
        # Find first position where value > threshold
        pos = bisect.bisect_right(entries, (threshold, chr(0x10FFFF)))
        return {e[1] for e in entries[pos:]}

    def gte(self, field: str, threshold: float) -> set[str]:
        entries = self._index.get(field, [])
        pos = bisect.bisect_left(entries, (threshold, ""))
        return {e[1] for e in entries[pos:]}

    def lt(self, field: str, threshold: float) -> set[str]:
        entries = self._index.get(field, [])
        pos = bisect.bisect_left(entries, (threshold, ""))
        return {e[1] for e in entries[:pos]}

    def lte(self, field: str, threshold: float) -> set[str]:
        entries = self._index.get(field, [])
        pos = bisect.bisect_right(entries, (threshold, chr(0x10FFFF)))
        return {e[1] for e in entries[:pos]}

    def get_sorted(self, field: str) -> list[tuple[float, str]]:
        return self._index.get(field, [])

    # ── Lifecycle ──────────────────────────────────────────────────────────────

    def clear(self) -> None:
        self._index.clear()

    @property
    def entry_count(self) -> int:
        return sum(len(v) for v in self._index.values())


def _to_number(value: Any) -> float | None:
    try:
        n = float(value)
        return n if (n == n) else None  # NaN check
    except (TypeError, ValueError):
        return None

File: rust/range.rs
// =============================================================================
// vapor-idx — indexes/range.rs
// Sorted numeric index using binary search.
// Supports: gt, lt, gte, lte
// =============================================================================

use std::collections::{HashMap, HashSet};

#[derive(Debug, Default)]
pub struct RangeIndex {
    // field → sorted Vec<(value_bits, id)>
    index: HashMap<String, Vec<(u64, String)>>,
}

impl RangeIndex {
    pub fn new() -> Self { Self::default() }

    pub fn add(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        let numbers: Vec<f64> = match value {
            serde_json::Value::Number(n) => n.as_f64().into_iter().collect(),
            serde_json::Value::Array(arr) => arr.iter()
                .filter_map(|v| v.as_f64())
                .collect(),
            _ => return,
        };

        let entries = self.index.entry(field.to_owned()).or_default();
        for num in numbers {
            let bits = num.to_bits();
            let pos = entries.partition_point(|(b, i)| b < &bits || (b == &bits && i.as_str() < id));
            entries.insert(pos, (bits, id.to_owned()));
        }
    }

    pub fn remove(&mut self, field: &str, value: &serde_json::Value, id: &str) {
        let numbers: Vec<f64> = match value {
            serde_json::Value::Number(n) => n.as_f64().into_iter().collect(),
            serde_json::Value::Array(arr) => arr.iter().filter_map(|v| v.as_f64()).collect(),
            _ => return,
        };
        let Some(entries) = self.index.get_mut(field) else { return };
        for num in numbers {
            let bits = num.to_bits();
            if let Ok(pos) = entries.binary_search(&(bits, id.to_owned())) {
                entries.remove(pos);
            }
        }
        if entries.is_empty() { self.index.remove(field); }
    }

    pub fn gt(&self, field: &str, threshold: f64) -> HashSet<String> {
        let entries = self.index.get(field).map(|v| v.as_slice()).unwrap_or(&[]);
        let bits = threshold.to_bits();
        // First position where value > threshold
        let pos = entries.partition_point(|(b, _)| *b <= bits);
        entries[pos..].iter().map(|(_, id)| id.clone()).collect()
    }

    pub fn gte(&self, field: &str, threshold: f64) -> HashSet<String> {
        let entries = self.index.get(field).map(|v| v.as_slice()).unwrap_or(&[]);
        let bits = threshold.to_bits();
        let pos = entries.partition_point(|(b, _)| *b < bits);
        entries[pos..].iter().map(|(_, id)| id.clone()).collect()
    }

    pub fn lt(&self, field: &str, threshold: f64) -> HashSet<String> {
        let entries = self.index.get(field).map(|v| v.as_slice()).unwrap_or(&[]);
        let bits = threshold.to_bits();
        let pos = entries.partition_point(|(b, _)| *b < bits);
        entries[..pos].iter().map(|(_, id)| id.clone()).collect()
    }

    pub fn lte(&self, field: &str, threshold: f64) -> HashSet<String> {
        let entries = self.index.get(field).map(|v| v.as_slice()).unwrap_or(&[]);
        let bits = threshold.to_bits();
        let pos = entries.partition_point(|(b, _)| *b <= bits);
        entries[..pos].iter().map(|(_, id)| id.clone()).collect()
    }

    pub fn clear(&mut self) { self.index.clear(); }

    pub fn entry_count(&self) -> usize {
        self.index.values().map(|v| v.len()).sum()
    }
}
